orderer/consensus/smartbft/consenter.go

```diff
@@ -0,0 +1,271 @@
+	configOptions := &smartbft.Options{}
+	consenters := support.SharedConfig().Consenters()
+	if err := proto.Unmarshal(support.SharedConfig().ConsensusMetadata(), configOptions); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal consensus metadata")
+	}
+
+	selfID, err := c.detectSelfID(consenters)
+	if err != nil {
+		return nil, errors.Wrap(err, "without a system channel, a follower should have been created")
+	}
+	c.Logger.Infof("Local consenter id is %d", selfID)
+
+	puller, err := newBlockPuller(support, c.ClusterDialer, c.Conf.General.Cluster, c.BCCSP)
+	if err != nil {
+		c.Logger.Panicf("Failed initializing block puller")
+	}
+
+	config, err := configFromMetadataOptions((uint64)(selfID), configOptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed parsing smartbft configuration")
+	}
+	c.Logger.Debugf("SmartBFT-Go config: %+v", config)
+
+	configValidator := &ConfigBlockValidator{
+		ChannelConfigTemplator: c.Registrar,
+		ValidatingChannel:      support.ChannelID(),
+		Filters:                c.Registrar,
+		ConfigUpdateProposer:   c.Registrar,
+		Logger:                 c.Logger,
+	}
+
+	chain, err := NewChain(configValidator, (uint64)(selfID), config, path.Join(c.WALBaseDir, support.ChannelID()), puller, c.Comm, c.SignerSerializer, c.GetPolicyManager(support.ChannelID()), support, c.Metrics, c.BCCSP)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed creating a new BFTChain")
+	}
+
+	return chain, nil
+}
+
+func (c *Consenter) IsChannelMember(joinBlock *cb.Block) (bool, error) {
+	if joinBlock == nil {
+		return false, errors.New("nil block")
+	}
+	envelopeConfig, err := protoutil.ExtractEnvelope(joinBlock, 0)
+	if err != nil {
+		return false, err
+	}
+	bundle, err := channelconfig.NewBundleFromEnvelope(envelopeConfig, c.BCCSP)
+	if err != nil {
+		return false, err
+	}
+	oc, exists := bundle.OrdererConfig()
+	if !exists {
+		return false, errors.New("no orderer config in bundle")
+	}
+	member := false
+	for _, consenter := range oc.Consenters() {
+		if bytes.Equal(c.Cert, consenter.ServerTlsCert) || bytes.Equal(c.Cert, consenter.ClientTlsCert) {
+			member = true
+			break
+		}
+	}
+
+	return member, nil
+}
+
+// TargetChannel extracts the channel from the given proto.Message.
+// Returns an empty string on failure.
+func (c *Consenter) TargetChannel(message proto.Message) string {
+	switch req := message.(type) {
+	case *ab.ConsensusRequest:
+		return req.Channel
+	case *ab.SubmitRequest:
+		return req.Channel
+	default:
+		return ""
+	}
+}
+
+func pemToDER(pemBytes []byte, id uint64, certType string, logger *flogging.FabricLogger) ([]byte, error) {
+	bl, _ := pem.Decode(pemBytes)
+	if bl == nil {
+		logger.Errorf("Rejecting PEM block of %s TLS cert for node %d, offending PEM is: %s", certType, id, string(pemBytes))
+		return nil, errors.Errorf("invalid PEM block")
+	}
+	return bl.Bytes, nil
+}
+
+func (c *Consenter) detectSelfID(consenters []*cb.Consenter) (uint32, error) {
```

Comment:
this is not how we should detect our self ID.

Reply:
Updated to use the ecert to detect the node id.